File: polynomial_regression.py
# Polynomial Regression

# Data Preprocessing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the Dataset
dataset = pd.read_csv('Position_Salaries.csv')
x = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values


# splitting the dataset into Training Set and Test Set
"""from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_x=StandardScaler()
x_train=sc_x.fit_transform(x_train)
x_test=sc_x.transform(x_test)"""

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(x, y)

# Analysing the result with the simple linear regression
"""from statsmodels.formula.api import OLS
regressor_1 = OLS(endog = y, exog = x).fit()
regressor_1.summary()"""

# Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
x_poly = poly_reg.fit_transform(x)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(x_poly, y)

# Analysing the result with the polynomial regression
"""regressor_2 = OLS(endog = y, exog = x_poly).fit()
regressor_2.summary()"""

# Visualising the Simple linear regression result
plt.scatter(x, y, color='red')
plt.plot(x, lin_reg.predict(x), color='blue')
plt.title('Truth OR Bluff (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Used for higher resolution and smoother curve
"""x_grid = np.arange(min(x), max(x), 0.1)
x_grid = x_grid.reshape(len(x_grid),1)"""

# Visualising the Polynomial regression result
plt.scatter(x, y, color='red')
plt.plot(x, lin_reg_2.predict(poly_reg.fit_transform(x)), color='blue') #Change x by x_grid if using 
plt.title('Truth OR Bluff (Ploynomial Regression)')                   #higher resolution and smoother curve
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Converting the float value to a matrix using numpy
i = 6.5
i = np.array(i).reshape(-1,1)

# Predicting a new result with Linear Regression
lin_reg.predict(i)

# Predicting a new result with Polynomial Regression
lin_reg_2.predict(poly_reg.fit_transform(i))

File: polynomial_regression.R
# Polynomial Regression
# Data Preprocessing

# Importing the Dataset

dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into tne Training set and test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)


# Feature Scaling
# training_set[ ,2:3] = scale(training_set[ ,2:3])
# test_set[ ,2:3] = scale(test_set[ ,2:3])

# Fitting Linear regression to the dataset
lin_reg = lm(formula = Salary ~ ., data = dataset)

# Fitting Polynomial regression to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ ., data = dataset)

# Visualising the Simple Linear Regression Result
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth OR Bluff (Linear Regression)') + 
  xlab('Position Level') + 
  ylab('Salary')

# Used for higher resolution and smoother curve
# x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

# Visualising the Polynomial Regression Result
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth OR Bluff (Polynomial Regression)') + 
  xlab('Position Level') + 
  ylab('Salary')

# Predicting a new result with Simple Linear Regression
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Predicting a new result with Polynomial Regression
y_pred = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                Level2 = 6.5^2,
                                                Level3 = 6.5^3,
                                                Level4 = 6.5^4 ))



